### ISSUE TITLE
Notes from my test drive of Pakyow

### ISSUE THREAD

**Issue by iamvery** [pakyow/pakyow#66]
I had a great time playing with your framework today! I'm quite compelled by the idea of separating all backend code from the views. I didn't get beyond the scope of the warmup exercise, but in the domain of that problem, things made a lot of sense and fell out quite elegantly.

While working with the warmup, I recorded some notes that I thought worth sharing. They cover a lot of different aspects so I thought it'd be better to drop them all in one place and allow you to act as appropriate. I'm not too bent out of shape about anything, so no need to rush. I just wanted to give back!

Please do let me know if you need any help or additional context. I'll do what I can!

---

# Thoughts on [Pakyow](http://pakyow.com)

## Tutorial Problems

* I don't think it was ever mentioned that `require 'twitter'` must be added to the helpers file to access that lib.
* I don't think it is mentioned that the user may need to stop/start the server when adding new gems to the app. That is sort of a "duh", but it's worth mentioning IMO to keep WATs at bay.
* When the `show` route is introduced it does give the user the code, but it doesn't specifically say where it belongs. My buffer was a little full for this to be immediately obvious to me at the time. Probably worth a pointer.
* I thought it was pretty sweet how the renderer knows to iterate over a collection to render multiple elements of scope. I'm not sure that was specifically mentioned.

## WATs

* I would have expected https://github.com/pakyow/pakyow/tree/master/pakyow-rake to depend on the rake gem
* The error message displayed when rendering a view with a scope that is missing is quite cryptic and unhelpful.
* I had to go ahead and fix up the "view details" link to be a link to this new view. That seems to have been necessary because once a route is defined for "show", the prototype is no longer available. I might have rather seen this prototype continue to work at this point. I wouldn't have necessarily expected the additional route to have the side effect of making this unaccessible (at least in dev mode)
* There seems to be a bug in pakyow prod on Heroku. See https://github.com/iamvery/pakyow-warmup/commit/7f7ae865f77aa20d931166fd5684819f10457040
* When the app experienced an error in twitter client auth on prod it reverted to what appeared to be "prototype" mode (rather than allowing the error to bubble up). This was quite a head-scratcher, as the app seemed to be working, just not "dynamic".

## Stylesheets

* It appears the built-in css doesn't load in prototype environment. Seems such an environment is specifically intended for views where styles are an integral aspect. My only thought is that it's only excluding "pakyow-css" styles which also appeared to happen in the prod environment.
* Speaking of which, I'm not sure I even understand "pakyow-css". The warmup indicates that it's included with generated apps, but it's not a dependency in graph. Maybe the css files included in the generation are from pakyow-css, but it doesn't specifically depend on that gem?

## Naming

* I don't particularly love the name of the `restful` routing method. IMO there is too much contention around the term REST for folks to agree with this. I really like using the term `resource` for an interface that represents a particular concept following "standard" CRUD operations. Maybe Rails got this one right? :tongue:
* Tab completion opening files got annoying with the `app.rb` file and `app` directory. "Meh", but worth mentioning.
* The concept of "bindings" in a pakyow app clashes with a "binding" in Ruby. I could see this being pretty weird in some places as bindings are such a core feature of the language.

## Organization

* It struck me as potentially messy to roll business logic into the same location as routing. That is given time on even a small project, this file would quickly grow unwhieldy. I did consider, however, that this may be intentionally simplified for the warmup leaving the task of extraction to the developer of a "real" app. I [took a stab](https://github.com/iamvery/pakyow-warmup/commit/0ed73072bd844efa23799a0a222c328ed968b2a6) at someting. While it does feel a little better to me, I doubt it's robust, and I'm not a huge fan of rolling the term "controller" into the app as it carries a lot of baggage.
* Similarly, it seems that helpers and bindings may also get out of hand. I [tried something](https://github.com/iamvery/pakyow-warmup/commit/e605f8accd722e92c52c72ec8f5b7c41bc1d5888) that seemed to be decent enough in isolating bindings.

**Comment by bryanp** [comment]
There's some really good input here! As I chew on it the next few days I'll be posting some thoughts :)

**Comment by bryanp** [comment]
Okay, it took a bit longer than I was expecting to ponder some of these things. I've tried to address most of the key points in your issue. Here we go.

# Tutorial Problems

These are all issues well worth fixing.

# WATs

I would also have expected `pakyow-rake` to depend on `rake`. Will fix this.

Error messages from the public API leave a lot to be desired in general. This is something we'll be cleaning up. Thanks for pointing out this particular example!

Re: the prototype no longer being available, I've been pondering this and don't have a great answer on it. I do think it's more an issue with how the warmup is structured than anything. I may try reorganizing the steps a bit.

The issue you hit on Heroku has been fixed for the next release.

# Stylesheets

I was unable to reproduce the issue of `pakyow-css` not loading in prototype mode. Can you provide some screenshots? It'd also be helpful to know if the assets are being requested and 404ing, or not being requested at all.

`pakyow-css` is just included with generated apps. It'll be more clear once we have a sane way to handle assets within a Pakyow app. The purpose of pakyow-css is to provide a lightweight, structure-only framework to whip up an app quickly and style it later without having to rip out the structural styling.

# Naming

I sort of agree with you on the `restful` vs `resource` thing. The `resource` concept is something we're working on as an even easier way of defining a REST API (though we may come up with a name other than `resource` since it won't be exactly like a resource in Rails). At some point the `restful` concept will be hidden from most users.

Excellent feedback on the `app.rb` naming. Might be worth renaming `app.rb` to `pakyow.rb` or something. Considering this!

Not sure if the bindings thing is something we'll change or not, but I am pondering it. It never crossed my mind about it conflicting with the concept of bindings in Ruby. However, I think it's only a mental conflict.

# Organization

We definitely want to avoid controllers in Pakyow. However, there is a way to break routes out into separate files:

```ruby
Pakyow::App.routes :post do
  restful :post, '/posts' do
    ...
  end
end
```

0.10 will add the concept of renderers, which abstracts the view logic out of the route definitions. By 1.0 we want the route definitions to just be glue between a view, the rendering logic, and data logic (a controller, basically).

You can also break bindings out into separate files just like routes. I hadn't thought before about addressing the helpers issue, but I am now considering a solution for that.

---

Let me know if you have any more input on this!

**Comment by bryanp** [comment]
@iamvery Ping! Just checking to see if you had any further comments before I close this out :-)

**Comment by iamvery** [comment]
I've had a long created task to respond to this. Reading back over it, I don't really have any other input. I'd love to try and reproduce the css quirk, but time is short and it doesn't feel particularly important. Hopefully I can check out the realtime stuff soon!

**Comment by bryanp** [comment]
Awesome, thanks again for taking the time to post this writeup! Many of the issues have been resolved for this week's 0.10 release and the others are on our radar.

**Comment by iamvery** [comment]
:tada:
